test/components/renku_data_services/solr/conftest.py

```diff
@@ -0,0 +1,124 @@
+from renku_data_services.solr.solr_client import SolrClientConfig
+from renku_data_services.solr.solr_migrate import SchemaMigrator
+
+
+@pytest.fixture(scope="session")
+def free_port() -> int:
+    lock = Lock()
+    with lock, socket.socket() as s:
+        s.bind(("", 0))
+        port = int(s.getsockname()[1])
+        return port
+
+
+@pytest.fixture(scope="session")
+def monkeysession():
+    mpatch = pytest.MonkeyPatch()
+    yield mpatch
+    mpatch.undo()
+
+
+@pytest.fixture(scope="session")
+def solr_bin_path():
+    solr_bin = getenv("SOLR_BIN_PATH")
+    if solr_bin is None:
+        solr_bin = "solr"
+    return solr_bin
+
+
+async def __wait_for_solr(host: str, port: int) -> None:
+    tries = 0
+    with httpx.Client() as c:
+        while True:
+            try:
+                c.get(f"http://{host}:{port}/solr")
+                return None
+            except Exception as err:
+                print(err)
+                if tries >= 20:
+                    raise Exception(f"Cannot connect to solr, gave up after {tries} tries.")
+                else:
+                    tries = tries + 1
+                    await sleep(1)
+
+
+@pytest_asyncio.fixture(scope="session")
+async def solr_instance(tmp_path_factory, free_port, monkeysession, solr_bin_path):
+    solr_root = tmp_path_factory.mktemp("solr")
+    solr_bin = solr_bin_path
+    port = free_port
+    logging.info(f"Starting SOLR at port {port}")
+    proc = subprocess.Popen(
+        [
+            solr_bin,
+            "start",
+            "-f",
+            "--host",
+            "localhost",
+            "--port",
+            f"{port}",
+            "-s",
+            f"{solr_root}",
+            "-t",
+            f"{solr_root}",
+        ],
+        env={"PATH": getenv("PATH", ""), "SOLR_LOGS_DIR": f"{solr_root}", "SOLR_ULIMIT_CHECKS": "false"},
+    )
+    monkeysession.setenv("SOLR_HOST", "localhost")
+    monkeysession.setenv("SOLR_PORT", f"{port}")
+    monkeysession.setenv("SOLR_ROOT_DIR", solr_root)
+
+    await __wait_for_solr("localhost", port)
+
+    yield
+    try:
+        proc.terminate()
+    except Exception as err:
+        logging.error(f"Encountered error when shutting down solr for testing {err}")
+        proc.kill()
+
+
+@pytest.fixture
+def solr_core(solr_instance, monkeysession):
+    core_name = str(ULID()).lower()
+    monkeysession.setenv("SOLR_CORE_NAME", core_name)
+    return core_name
+
+
+@pytest.fixture()
+def solr_config(solr_core, solr_bin_path):
+    core = solr_core
+    solr_host = getenv("SOLR_HOST")
+    solr_port = getenv("SOLR_PORT")
+    solr_config = SolrClientConfig(base_url=f"http://{solr_host}:{solr_port}", core=core)
+    solr_bin = solr_bin_path
+    subprocess.run([solr_bin, "create", "-c", core])
+
+    # TODO this is annoying!
+    dir = getenv("SOLR_ROOT_DIR")
+    conf_file = f"{dir}/{core}/conf/managed-schema.xml"
+    subprocess.run(["chmod", "644", conf_file])
```

Comment:
suggestion: If the solr executable can run with a core located in an arbitrary location, then pytest can give you a fixture that will be a location in the /tmp or some other folder (it tries to do the right thing for the OS). Then you can put the file there and the file will be deleted when not needed. See https://docs.pytest.org/en/stable/how-to/tmp_path.html

Alternatively we can also just create and commit this file in the repo and use it just for tests. Then we can also set the permissions on it.

Reply:
This is also a bit strange. The `SOLR_ROOT_DIR` is actually created by pytest (as you suggested) and it is set into the session as an env var (one with `scope=session`). This fixture here, then creates a new core at solr. So it runs for every test. Creating a core results in solr creating some directories (with the random core name). But for some reason (sigh), they are created with `444` permission. That makes all the schema commands error out, because solr can't update this file… (I'm not sure why it writes into it, tbh). This is why this fix/hack makes the file writable.

Reply:
Aha ok. Cool. Then a minor edit to what you have above. Instead of `subprocess.run` use:

```python
import os
import stat
os.chmod(conf_file, stat.S_IRUSR | stat.S_IWUSR | stat.S_IROTH | stat.S_IRGRP)
```

it is just nicer to now shell out. But it is not a big deal either way.